caddy/setup/browse_test.go

```diff
@@ -0,0 +1,77 @@
+package setup
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/mholt/caddy/caddy/parse"
+	"github.com/mholt/caddy/middleware/browse"
+	"github.com/mholt/caddy/server"
+)
+
+func TestBrowse(t *testing.T) {
+
+	tempDirPath, err := getTempDirPath()
+	if err != nil {
+		t.Fatalf("BeforeTest: Failed to find an existing directory for testing! Error was: %v", err)
+	}
+	nonExistantDirPath := filepath.Join(tempDirPath, strconv.Itoa(int(time.Now().UnixNano())))
+
+	tempTemplate, err := ioutil.TempFile(".", "tempTemplate")
+	if err != nil {
+		t.Fatalf("BeforeTest: Failed to create a temporary file in the working directory! Error was: %v", err)
+	}
+	defer os.Remove(tempTemplate.Name())
+
+	tempTemplatePath := filepath.Join(".", tempTemplate.Name())
+
+	testTokens := []string{
+		"browse " + tempDirPath + "\n browse .",
+		"browse /",
+		"browse . " + tempTemplatePath,
+		"browse . " + nonExistantDirPath,
+		"browse " + tempDirPath + "\n browse " + tempDirPath,
+	}
+
+	tests := []struct {
+		expectedPathScope []string
+		shouldErr         bool
+	}{
```

Comment:
How about making the input another field in this struct? In other words, combine `testTokens` and `tests`.

Then you can actually iterate an anonymous slice below instead, like this:

```go
for i, test := range []struct{
    // input, expectedPathScope, shouldErr fields here
}{
    // ... test cases here ...
}{
    // test logic here
}
```

Can use a similar strategy for redir_test.go too.